sdk/eventhub/azure-eventhubs/azure/eventhub/aio/producer_async.py

```diff
@@ -213,12 +216,26 @@ def _on_outcome(self, outcome, condition):
                 :caption: Sends an event data and blocks until acknowledgement is received or operation times out.
 
         """
+        # Tracing code
+        span_impl_type = settings.tracing_implementation()  # type: Type[AbstractSpan]
+        if span_impl_type is not None:
+            child = span_impl_type(name="Azure.EventHubs.send")
+            child.kind = SpanKind.CLIENT  # Should be PRODUCER
+
+        def trace_message(message):
+            message_span = child.span(name="Azure.EventHubs.message")
+            message_span.start()
+            app_prop = dict(message.application_properties)
+            app_prop.setdefault(b"Diagnostic-Id", message_span.get_trace_parent().encode('ascii'))
```

Comment:
[setdefault](https://docs.python.org/3/library/stdtypes.html#dict.setdefault) will not owerwrite the item if it is already set. We could avoid the cost of encoding by doing an `in` check, however.

Reply:
I don't expect this encode to take a lot, since the string is 0-9A-Z and the encoder is ascii. That will let mostly cast it "as-is" under the hood. 
With the `in`, in case we do need to assign it, we will do another lookup to find the right spot. Without the `in`, we indeed ascii-encode for nothing sometimes, but if we need to insert we know already it is.